src/python/pants/backend/python/rules/hermetic_pex.py

```diff
@@ -1,34 +1,161 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-from typing import Any, Iterable, Mapping, Optional
+import os
+from dataclasses import dataclass
+from typing import Any, Iterable, Mapping, Optional, Tuple
 
 from pants.backend.python.subsystems.subprocess_environment import SubprocessEnvironment
+from pants.engine import process
+from pants.engine.engine_aware import EngineAware
 from pants.engine.fs import Digest
-from pants.engine.process import Process
+from pants.engine.process import BinaryPathRequest, BinaryPaths, Process
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.option.subsystem import Subsystem
 from pants.python.python_setup import PythonSetup
+from pants.util.logging import LogLevel
+from pants.util.memo import memoized_property
+from pants.util.ordered_set import OrderedSet
 from pants.util.strutil import create_path_env_var
 
 
+@dataclass(frozen=True)
+class PexEnvironment(EngineAware):
+    path: Iterable[str]
+    interpreter_search_paths: Iterable[str]
+    bootstrap_python: Optional[str] = None
+
+    def create_argv(self, pex_path: str, *args: str) -> Iterable[str]:
+        argv = [self.bootstrap_python] if self.bootstrap_python else []
+        argv.extend((pex_path, *args))
+        return argv
+
+    @property
+    def environment_dict(self) -> Mapping[str, str]:
+        return dict(
+            PATH=create_path_env_var(self.path),
+            PEX_PYTHON_PATH=create_path_env_var(self.interpreter_search_paths),
+        )
+
+    def level(self) -> Optional[LogLevel]:
+        return LogLevel.INFO if self.bootstrap_python else LogLevel.WARN
+
+    def message(self) -> Optional[str]:
+        if not self.bootstrap_python:
+            return (
+                "No bootstrap python executable could be found. "
+                "Will attempt to run PEXes directly."
+            )
+        return f"Selected {self.bootstrap_python} to bootstrap PEXes with."
+
+
+class PexRuntimeEnvironment(Subsystem):
+    """How Pants uses Pex to run Python subprocesses."""
+
+    options_scope = "pex"
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--binary-search-path",
+            advanced=True,
+            type=list,
+            default=["<PATH>"],
+            metavar="<binary-paths>",
+            help=(
+                "A list of paths to search for binaries needed by PEX subprocess. The special "
+                'string "<PATH>" will expand to the contents of the PATH env var.'
+            ),
+        )
+        register(
+            "--bootstrap-interpreter-names",
+            advanced=True,
+            type=list,
+            default=["python", "python3", "python2"],
+            metavar="<bootstrap-python-names>",
+            help=(
+                "The names of python binaries to search for to bootstrap PEX files with. Earlier "
```

Comment:
An important thing to highlight here (assuming that it is true?) is that this has no influence on which interpreter ends up being used to actually run your code. The rest can probably just stay in docstrings (especially since we render a WARN currently if the lookup fails).

Reply:
Good suggestion. Thanks.